draft-ietf-jsonpath-base.md

```diff
@@ -1502,10 +1502,24 @@ The well-typedness of function expressions can now be defined in terms of this t
 
 For a function expression to be well typed:
 
-1. its declared type must be well typed in the context in which it occurs, and
-2. its arguments must be well typed for the declared type of the corresponding parameters.
+1. its function name MUST be registered (see {{iana-fnex}}),
+2. its declared type must be well typed in the context in which it occurs, and
+3. its arguments must be well typed for the declared type of the corresponding parameters.
 
-(1) As per the grammar, a function expression can occur in three different
+(1) The registered set of function extensions is the source of knowledge regarding the declared return and
+parameter types of a function extension.
+
+If a function expression is treated as well typed, the function extensions involved in the expression belong
+to the registered set.
+However, an implementation may treat a function expression as not well typed even if all the function extensions
+involved in the expression are registered.
+Implementations MUST support the function extensions defined
+in this document, but are free to support or not support other registered function extensions.
+
+If an implementation does not support a function extension used in a query, it MUST treat
+any function expression involving the function extension as not well typed.
```

Comment:
I still don't understand this association between registration and well-typedness.

If a function isn't recognized, then _that_ is the error.  The implementation cannot determine the well-typedness of a function it doesn't know.  It may be that the expression _is_ well-typed, but the implementation just doesn't support the function.  In this case, raising an error saying that it's not well-typed is incorrect.

Reply:
Section 2.1 describes the high level error processing model: if a query is not well formed (conforms to the ABNF) and valid, an error MUST be raised (before the argument is considered). Validity (of a well formed query) means fulfilling all the semantic requirements of the spec: integers are in range and uses of function extensions are well typed.

We _could_ introduce a third category of semantic requirement to cover registration, but that's not necessary and it would introduce more function extension detail into section 2.1 which I think would be a bit of a distraction at that stage in the document.

As you point out, if an implementation can't determine the result type of a function extension, it can't know that an expression involving the function extension is well typed. I am proposing that in such a situation, the implementation should presume that the expression is not well typed. (Any error message can certainly point out that the function extension isn't recognised, and I'd expect helpful implementations to do that.)

There is a parallel in distributed two phase commit which may be helpful. When the global outcome of a transaction is (permanently) in doubt, there are three policies for handling the situation. "Presumed abort" presumes the transaction failed. "Presumed commit" presumes the transaction succeeded. "Presumed nothing" makes no presumption: it essentially waits until the global outcome can be determined (or at least asserted, e.g. by a human operator). The parallel is that I am proposing that when the well typedness of a function expression is not known, we should presume it is not well typed. The policy of presuming the function expression is well typed is non-sensical. I _think_ you are suggesting that it would be better not to make a presumption, _but to fail with an error anyway_, which presumes invalidity (of a different kind).

Reply:
> Any error message can certainly point out that the function extension isn't recognised, and I'd expect helpful implementations to do that.

If this is your expectation, then why require that the error is "not well-typed"?  Either the spec is going to dictate errors or it's not.

Reply:
> when the well typedness of a function expression is not known, we should presume it is not well typed.

We don't have enough information.  Making this presumption is wrong, and requiring implementations to do so, is doubly wrong.

Reply:
> > Any error message can certainly point out that the function extension isn't recognised, and I'd expect helpful implementations to do that.
> 
> If this is your expectation, then why require that the error is "not well-typed"? Either the spec is going to dictate errors or it's not.

There is no requirement that the error is "not well-typed". The only requirement is that the query fails with an error. This is consistent with the policy in section 2.1.

Reply:
> > when the well typedness of a function expression is not known, we should presume it is not well typed.
> 
> We don't have enough information. Making this presumption is wrong, and requiring implementations to do so, is doubly wrong.
> 
> Implementations should error, yes. And they should do so at the parsing stage. But an unrecognized function is just that: unrecognized; and the spec has no business saying the error should suggest anything else.

This PR does not assume there will be an error message along the lines of "not well-typed". An implementation is free to produce an error message along the lines "error: function extension foo() is not supported".

Remember my fundamental reason for adopting this policy is to avoid raising too many details about function extensions to the level of section 2.1 (and perhaps elsewhere in the spec).

Reply:
> This PR does not assume there will be an error message along the lines of "not well-typed".

I think Greg and I are reading "treat as not well-typed" as saying just that.  I think we are looking for some other way of saying that without that implication, e.g., "cannot treat as well-typed".

Reply:
Ah, since I may be in a minority of one, I'll try adjusting the approach.

Reply:
@gregsdennis and @cabo: please see the update. Section 2.1 is impacted slightly, but perhaps this is worth it to cleanly separate supportedness from well-typedness.

Reply:
It may be worth keeping in mind that is it much easier to update an implementation with new registry information than to actually implement the functions (in particular if they rely on platform capabilities you may not have).
So implementations may be able to provide accurate information about welltypedness of a query without being able to actually execute it -- we need to keep these two separate.

Reply:
This leads to the observation that there is an objective and a subjective view of validity.  A query may be valid objectively (in a god-view, who always has the most recent registration updates), it may also be valid subjectively (which should only be true if it is also valid objectively), or it may be of an unknown validity subjectively (cannot happen from an objective point of view).

Reply:
@cabo & @gregsdennis: I made another change which makes supportedness and (subjective) registration of a function extension othogonal.

Let's consider the four combinations for a function extension `foo()`:

* supported and registered - no error
* unsupported and registered - error "foo() unsupported" (see §2.4)
* supported and unregistered - error A (see below)
* unsupported and unregistered - either error "foo() unsupported" (see §2.4) or error A (see below), at the choice of the implementation

What should error A be for an unregistered function extension? Since well-typedness of the containing function expression cannot be known, we could raise an error "function expression involving foo() not necessarily well typed"). Alternatively, we could separate unregistration from well-typedness and raise an error "unregistered foo()" without saying anything about well-typedness.

If we separated unregistration from well-typeness, we could either make unregistration its own kind of invalidity (i.e. a peer of unsupported and not well typed) or we could collapse unregistration and unsupported into a single kind of invalidity (essentially saying that an implementation cannot support a function extension that is unregistered, at least not in strict/spec-compliant mode).

I didn't separate subjective and objective registration in the above. Perhaps that needs doing. Subjective registration seems more interesting to implementers while objective registration may be more interesting to users.

Any thoughts on the best way forward?

Reply:
> essentially saying that an implementation cannot support a function extension that is unregistered, at least not in strict/spec-compliant mode

I think this is the correct approach.  An implementation would need to explicitly "unsupport" any custom functions when operating in strict mode.  Further, that strict mode should be the default.

Regarding subjective vs objective support, the implementation supports what it supports.  A user can't realistically expect an implementation to support a function that was added to the registry after the library was released, for example.  The library should document what functions it supports (just like any other functionality), and the user should be expected to read such documentation.  (Strangely, I'm having this same conversation in JSON Schema Land.)